pulp_2to3_migration/app/plugin/deb/pulp_2to3_models.py

```diff
@@ -184,3 +184,257 @@ def create_pulp3_content(self):
+                    architecture=architecture,
+                    distribution=component.distribution,
+                    codename=component.release,
+                    suite=release.suite,
+                    pulp2content=pulp2_unit_map[component.id],
+                ))
+
+        cls.objects.bulk_create(
+            component_units_to_save,
+            ignore_conflicts=True,
+            batch_size=DEFAULT_BATCH_SIZE
+        )
+        Pulp2DebComponentPackage.objects.bulk_create(
+            component_package_units_to_save,
+            ignore_conflicts=True,
+            batch_size=DEFAULT_BATCH_SIZE
+        )
+        Pulp2DebReleaseArchitecture.objects.bulk_create(
+            release_architecture_units_to_save,
+            ignore_conflicts=True,
+            batch_size=DEFAULT_BATCH_SIZE
+        )
+
+    def create_pulp3_content(self):
+        """
+        Create a Pulp 3 detail Content unit for saving it later in a bulk operation.
+
+        Return an unsaved Pulp 3 Content
+        """
+        release = pulp3_models.Release.objects.filter(
+            distribution=self.distribution,
+            codename=self.codename,
+            suite=self.suite,
+        ).first()
+        pulp3_component = pulp3_models.ReleaseComponent(
+            component=self.component,
+            release=release,
+        )
+        return (pulp3_component, None)
+
+
+class Pulp2DebComponentPackage(Pulp2to3Content):
+    """
+    Pulp 2to3 detail content model to store pulp 2 DebComponent content details for Pulp 3
+    content creation.
+    """
+    pulp2_type = 'deb_component'
+    package_relative_path = django_models.TextField()
+    package_sha256 = django_models.TextField()
+    component = django_models.TextField()
+    distribution = django_models.TextField()
+    codename = django_models.TextField()
+    suite = django_models.TextField()
+
+    class Meta:
+        default_related_name = 'deb_component_package_detail_model'
+        unique_together = (
+            'package_relative_path',
+            'package_sha256',
+            'component',
+            'distribution',
+            'codename',
+            'suite',
+        )
+
+    @classmethod
+    def pre_migrate_content_detail(cls, content_batch):
+        """
+        Pre-migrate Pulp 2 content with all the fields needed to create a Pulp 3 Content
+
+        Args:
+             content_batch(list of Pulp2Content): pre-migrated generic data for Pulp 2 content.
+        """
+        pass
+
+    def create_pulp3_content(self):
+        """
+        Create a Pulp 3 detail Content unit for saving it later in a bulk operation.
+
+        Return an unsaved Pulp 3 Content
+        """
+        release_component = pulp3_models.ReleaseComponent.objects.filter(
+            component=self.component,
+            release__distribution=self.distribution,
+            release__codename=self.codename,
+            release__suite=self.suite,
+        ).first()
```

Comment:
In the current implementation retrieving this `release_component` (for associating it) appears to fail, because it was never saved to the data base. However, if I comment out or remove lines 50 and 51 the pipeline does finish successfully. Therefore I know that creating release components does work in principle, and I also know that retrieving units for association does work in principle (since creating release components also does this).

I can only conclude that this is some order issue, where python is trying to retrieve the release_component before it is saved to the DB. However, I was under the impression that content types are migrated in turn, and thus should guarantee order. (Otherwise what is the point of using an ordered dict in line 46 and the comment found [here](https://github.com/pulp/pulp-2to3-migration/blob/ed3fdc3457dce5fc1981acd26f779ef16a587944/pulp_2to3_migration/app/plugin/content.py#L178)?)

I have been stuck on this for a while now, and any help would be greatly appreciated.
It might be easier to explain in a screen session, so I would always be up for one of those.

Reply:
I'm struggling with one thing, please let me know where my understanding/logic is off.
`pre_migrate_content_detail` does nothing, it means that all values for this class (Pulp2DebComponentPackage) are null or empty string, correct? If so, then do you expect a ReleaseComponent with all empty values, e.g. self.distribution, self.component, etc?

And what are the lines 50/51 do you refer to? in which module? Thanks!

Reply:
Wups, I am referring to lines 50, 51 in `pulp_2to3_migration/app/plugin/deb/migrator.py`.

The fact that `pre_migrated_content_detail` is empty here is a bit of a dirty workaround around the plugin design.
I basically need to create several pulp 3 types from a single pulp 2 type. To avoid having to retrieve the pulp2 units from mongo multiple times, I made the `pre_migrate_content_detail` from `Pulp2DebComponent` create not only `Pulp2DebComponent` objects, but also `Pulp2DebComponentPackage` and `Pulp2DebReleaseArchitecture` objects.
The `pre_migrate_content_detail` functions from the latter two classes then no longer need to do anything. Nevertheless all the pre migrated content I need is created correctly.

Reply:
Now that I read that, I guess that does not actually avoid batches from being retrieved from mongo and passed to the functions that then don't do anything... Maybe I should just split up the `pre_migrated_content_detail` function from `Pulp2DebComponent` into three.

Reply:
I have now updated the docstrings to make clear why some `pre_migrate_content_detail` methods don't do anything.
My original issue was fixed by adding the relevant types to the `future_types` list in the migrator.

Resolved.